File: ioi2020-mushrooms/solution/parallel-one-by-one-divnconq.py
#
# A parallel implementation of the one-by-one solution (creating threads through divide and conquer).
# The purpose of the parallel solutions is to verify if the graders have the correct behavior
#   when contestants use parallelism in their solutions.
#
# Author: Kian Mirjalali
#
from typing import Final, List
from threading import Thread
from stub import use_machine

def count_mushrooms(n: int) -> int:
    species: Final[List[int]] = n * [0]
    max_threads: Final[int] = 200
    block_size: Final[int] = max(1, n//max_threads)
    def handle_range(begin: int, end: int):
        l: Final[int] = end - begin
        if l <= 0:
            return
        if l <= block_size:
            for i in range(begin, end):
                species[i] = use_machine([0, i])
            return
        # l >= 2
        mid: Final[int] = begin + l//2
        th: Thread = Thread(target=handle_range, args=(begin, mid))
        th.start()
        handle_range(mid, end)
        th.join()
    handle_range(1, n)
    return species.count(0)

File: ioi2020-mushrooms/solution/parallel-one-by-one-divnconq.java
/*
 * A parallel implementation of the one-by-one solution (creating threads through divide and conquer).
 * The purpose of the parallel solutions is to verify if the graders have the correct behavior
 *   when contestants use parallelism in their solutions.
 *
 * Author: Kian Mirjalali
 */
public class mushrooms {

	private int use_machine(int[] x) {
		return stub.use_machine(x);
	}

	public int count_mushrooms(int n) {
		final int[] species = new int[n];
		species[0] = 0;
		final int max_threads = 200;
		final int block_size = Math.max(1, n/max_threads);
		new Object() {
			public void handle_range(int begin, int end) {
				final int l = end - begin;
				if (l <= 0)
					return;
				if (l <= block_size) {
					for (int i = begin; i < end; i++)
						species[i] = use_machine(new int[] {0, i});
					return;
				}
				// l >= 2
				final int mid = begin + l/2;
				Thread th = new Thread(()-> { handle_range(begin, mid); });
				th.start();
				handle_range(mid, end);
				try {
					th.join();
				} catch(InterruptedException e) {
					th.interrupt();
				}
			}
		}.handle_range(1, n);
		return (int)(java.util.Arrays.stream(species).filter(t -> (t == 0)).count());
	}
}
